### PR TITLE
perf(python): speed up `in series` 10x

### PR DESCRIPTION
**Pull request by jonashaag**
refs #9793

### CODE REVIEW

**Comment by jonashaag** [comment]
Hmm, this is 10x faster but this implementation is 1000x faster:


```
return item in self.to_numpy()
```

**Comment by jonashaag** [comment]
Even faster (3500x)!

```
return item in self.to_pandas(use_pyarrow_extension_array=False)
```

File: py-polars/polars/series/series.py [hunk 1]
```diff
@@ -869,6 +869,12 @@ def __copy__(self) -> Self:
     def __deepcopy__(self, memo: None = None) -> Self:
         return self.clone()
 
+    def __contains__(self, item) -> bool:
+        try:
```
Comment (ritchie46) [inline py-polars/polars/series/series.py:873]: Can you add a 

```python
# `TODO! optimize via `is_in` and `SORTED` flags`
```

**Review by ritchie46 — COMMENTED** [review]
Thanks! I have one comment as I think we should do this differently in the future. Can you add that comment?

**Comment by jonashaag** [comment]
Sure! I also tried `is_in` but it is way slower. I think it is because it will build a set first and in the particular case I benchmarked all values are unique.

I wonder if we should try to optimize `is_in` for cases where the haystack is sorted or something like that.

File: py-polars/polars/series/series.py (hunk 1 above)
Comment (jonashaag) [reply py-polars/polars/series/series.py:873]: Btw do you know there is a special feature in GitHub for concrete change suggestions in review comments?

**Comment by MarcoGorelli** [comment]
> Even faster (3500x)!
> 
> ```
> return item in self.to_pandas(use_pyarrow_extension_array=False)
> ```

that's because it's doing something different - it's checking if it's in the index, which is a `RangeIndex` in this case. So it's https://github.com/pola-rs/polars/issues/9793#issuecomment-1627808681 again

```python
In [7]: ser = pd.Series([1,2,3])

In [8]: ser
Out[8]:
0    1
1    2
2    3
dtype: int64

In [9]: 3 in ser
Out[9]: False
```

**Comment by MarcoGorelli** [comment]
😆 no worries, and I agree that (like most index operations) it's super-unintuitive (but unfortunately too late and hard to change)

anyway, if you type `item` as `Any` and run `mypy` to check it passes, this should be good

**Review by MarcoGorelli — APPROVED** [review]
Nice one, looks like this preserves the current behaviour and speeds things up even in the "best case":
```python
In [50]: pldf = pl.Series([0.]*1_000_000).to_frame().with_row_count()

In [51]: %timeit (pldf['row_nr'] == 0).any()
847 µs ± 20.8 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)

In [52]: %timeit 0 in pldf['row_nr']
888 µs ± 8.28 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
```

A separate topic might be whether or not `incompatible_value in series` should raise (like `series == incompatible_value` does), but that's a separate

Last thing (I think) is to just put `perf(python) in the PR title`

Leaving open some time in case Ritchie has other comments

**Comment by jonashaag** [comment]
> A separate topic might be whether or not `incompatible_value in series` should raise

I copied Pandas' behaviour here which doesn't raise

**Comment by jonashaag** [comment]
What do you think about going the faster `.to_numpy()` route?

**Comment by MarcoGorelli** [comment]
numpy isn't a required dependency in polars so I don't think that'd be viable

**Comment by jonashaag** [comment]
Actually it's slower with NumPy, I used a dev build of Polars for benchmarking, shame on me

**Comment by alexander-beedie** [comment]
> Actually it's slower with NumPy, I used a dev build of Polars for benchmarking, shame on me

Another common trap - I have done the same... ;)

**Pull request merged**